index.js

```diff
@@ -0,0 +1,124 @@
+class ConcurrencyQueue {
+
+  constructor(concurrency = 0) {
+    this.concurrency = concurrency;
+    this.queue = [];
+    this.count = 0;
+    this.stopped = true;
+  }
+
+  start() {
+    if (this.stopped) {
+      this.stopped = false;
+      this.run();
+    }
+  }
+
+  async push(task) {
+    this.queue.push(task);
+    this.start();
+  }
+
+  runTask() {
+    const task = this.queue.shift();
```

Comment:
before running next tasks you have to check if the queue is empty, or if it is stopped.